### PR TITLE
Support for inspecting Immutable.js objects

### PR DESCRIPTION
**Pull request by garetht**
This pull request makes changes meant to support the functionality requested for in https://github.com/facebook/react-devtools/issues/52.

**Changes Made**
- When an object is an Immutable object, it is now expandable so that its properties can be examined
- The display name for the object notes that it is an Immutable object
- Additional tests for the dehydrator and Immutable utilities have been added

### CODE REVIEW

**Comment by jaredly** [comment]
Hmmm. So that looks reasonable.
My inclination would be to instead implement this as a plugin, and either:
1 - make it so that plugins can alter the representation of props, or
2 - show immutable objects in an extra section in the sidebar, separate from props

Thanks for putting this together!
@spicyj do you have an opinion on how this is implemented?

**Comment by garetht** [comment]
That makes sense to me. Do you have a sample plugin I can consult?

File: agent/immutableUtils.js [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+          name += types[i];
+          break;
+        }
+      }
+      return name;
+    } else {
+      return data.constructor.name;
+    }
+  },
+  shallowToJS(dataStructure: any): any {
```
Comment (jackwanders) [inline agent/immutableUtils.js]: May I ask your rationale behind providing a `shallowToJS` here but not a `deepToJS` that calls [`Iterable.toJS()`](https://facebook.github.io/immutable-js/docs/#/Iterable/toJS)? For example, if a particular prop is an `Immutable.KeyedIterable` with a property value that is itself an Immutable object, then it seems that `shallowToJS` would not convert the property value to a mutable object.

Comment (garetht) [inline agent/immutableUtils.js]: Was just about to write that. The reason in edit 2 is precisely why.

**Comment by jackwanders** [comment]
@garetht, I believe @jaredly is referring to the idea of `extraPanes`, as implemented in [`frontend/Panel.js`](https://github.com/facebook/react-devtools/blob/3948128bf12696e9c2dc0fd20c84dc8ffa5b30c6/frontend/Panel.js#L231). Currently, it looks like the `rnStyle` capability is used to conditionally add an extra pane for `ReactNative` styles.

I'm not sure what the most appropriate way to implement these utils as a conditional functionality is, but one option might be to render an `ImmutableInspector` pane underneath the existing component inspection pane when the user inspects an immutable object. Some kind of top level toggle would be nicer, in my opinion, but may require more work as there doesn't appear to be an existing UI element in which to render something like a checkbox that, when checked, calls `.shallowToJS` on all immutable objects.

**Comment by jackwanders** [comment]
@garetht, @jaredly, I took a pass at implementing the "inspect immutable objects as js" functionality as an optional setting.

https://github.com/garetht/react-devtools/pull/1

This pull request creates a new checkbox at the top of `PropState` which, when checked, will toggle a property in `Store` and force an update of `Container`. The change wound up touching more modules than I had hoped, and someone with a better understanding of the devtools architecture may be able to clean things up.

**Comment by jaredly** [comment]
Yeah -- I'm hesitant to bake in Immutable support; it would be awesome to see this as a plugin; I'll take some time this weekend to figure out what that would look like

**Comment by Daniel15** [comment]
:+1:  I'm looking forward to this!

**Comment by zackargyle** [comment]
I decided against using Immutable for this very reason. It really renders the dev tools useless in inspecting data. As long as immutability is being encouraged through projects like redux, I think the few extra lines of code should definitely be baked into the devtools.

**Comment by kristian-puccio** [comment]
You can always select an object in the dev tools and query it using $r I
think. Then just use .toJS() in the console to print out the value.

But yeah it's definitely not as quick as it could be

**Comment by Daniel15** [comment]
@zackargyle - You could always clone @garetht's version of the React devtools (ie. the branch this pull request points to) and try it out :)

**Comment by gaearon** [comment]
As for the feature, I think it is really necessary we get something like this merged.
However I’m not sure about this implementation.

I think we should strive to support any iterables, and not just Immutable.js specifically. As an example, [react-json-tree](https://github.com/alexkuz/react-json-tree) checks whether [something has an iterator](https://github.com/alexkuz/react-json-tree/blob/7b1ccfa1e641ca2005afb2e6968f5ead114b5db0/src/objType.js#L3) and [handles iterable nodes in a specific way](https://github.com/alexkuz/react-json-tree/blob/7b1ccfa1e641ca2005afb2e6968f5ead114b5db0/src/JSONIterableNode.js). I think we should do the same here.

Would you like to look into reworking this PR to support arbitrary iterables? Thanks.

**Comment by gaearon** [comment]
Closing as there is no response from the author.

**Comment by gaearon** [comment]
A more generic support for this landed in https://github.com/facebook/react-devtools/pull/459.

**Pull request closed without merging**